### PR TITLE
ENH: Better error message for axis=None in `np.put_along_axis` and `np.take_along_axis`

### LINKED ISSUE #26553: ENH: Better error message for `np.put_along_axis`
### Proposed new feature or change:

In the following code snippet and its corresponding traceback,
```py
arr = np.array([[10, 30, 20], [60, 40, 50]])
indices = np.array([[1], [0]])
values = np.array([[-1], [-2]])

np.put_along_axis(arr, indices, values, axis=None)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[43], line 5
      2 indices = np.array([[1], [0]])
      3 values = np.array([[-1], [-2]])
----> 5 np.put_along_axis(arr, indices, values, axis=None)

File ~/Applications/miniconda3/envs/np/lib/python3.10/site-packages/numpy/lib/shape_base.py:260, in put_along_axis(arr, indices, values, axis)
    257     arr_shape = arr.shape
    259 # use the fancy index
--> 260 arr[_make_along_axis_idx(arr_shape, indices, axis)] = values

File ~/Applications/miniconda3/envs/np/lib/python3.10/site-packages/numpy/lib/shape_base.py:32, in _make_along_axis_idx(arr_shape, indices, axis)
     30     raise IndexError('`indices` must be an integer array')
     31 if len(arr_shape) != indices.ndim:
---> 32     raise ValueError(
     33         "`indices` and `arr` must have the same number of dimensions")
     34 shape_ones = (1,) * indices.ndim
     35 dest_dims = list(range(axis)) + [None] + list(range(axis+1, indices.ndim))

ValueError: `indices` and `arr` must have the same number of dimensions
```

the `indices` in the `ValueError` refers to the fancy index (`_make_along_axis_idx(arr_shape, indices, axis)`) but it causes confusion as it can also mean the `indices` passed by the user for the `indices` arg in `np.put_along_axis`. Nevertheless, it does not throw a clear error why the input arguments passed are invalid.

One partial solution is to rename `indices` of `_make_along_axis_idx` as `fancy_index` but that in itself does not explain what's causing the error. Or else, we can also possibly add an additional error check for the case where axis is None here:
https://github.com/numpy/numpy/blob/5ac5c15e8736ecb687dff3748aec7b3d57e81e1b/numpy/lib/_shape_base_impl.py#L254

### PR DESCRIPTION
**Pull request by arunppsg**
Fixes #26553 by adding a conditional check on indices for cases
where axis is None.

### CODE REVIEW

**Comment by mattip** [comment]
Please add a test or modify an existing test with extra cases to check the appropriate message is emitted. For `take_along_axis`, I think you can improve the tests [here](https://github.com/numpy/numpy/blob/c41621b3bb7c5e387049bab25794090635a2faa4/numpy/lib/tests/test_shape_base.py#L58), and for `put_along_axis` tests can be added [here](https://github.com/numpy/numpy/blob/c41621b3bb7c5e387049bab25794090635a2faa4/numpy/lib/tests/test_shape_base.py#L83)

File: numpy/lib/tests/test_shape_base.py [hunk 1]
```diff
@@ -104,6 +106,21 @@ def test_broadcast(self):
...
 
+    def test_invalid(self):
+        """ Test invalid inputs """
+        a_base = np.array([[10, 30, 20], [60, 40, 50]])
+        indices = np.array([[0], [1]])
+        values = np.array([[2], [1]])
+
+        # sanity check
+        a = a_base.copy()
+        put_along_axis(a, indices, values, axis=0)
```
Comment (mattip) [inline numpy/lib/tests/test_shape_base.py:117]:
```suggestion
        put_along_axis(a, indices, values, axis=0)
        assert np.all(a == [[2, 2, 2], [1, 1, 1]])
```

File: numpy/lib/_shape_base_impl.py [hunk 2]
```diff
@@ -252,6 +255,9 @@ def put_along_axis(arr, indices, values, axis):
     """
     # normalize inputs
     if axis is None:
+        if indices.ndim != 1:
+            raise ValueError(
+                'for axis=None, `indices` should have a dimension of one.')
```
Comment (mattip) [inline numpy/lib/_shape_base_impl.py:260]:
```suggestion
                'when axis=None, `indices` must have a single dimension.')
```

File: numpy/lib/_shape_base_impl.py [hunk 3]
```diff
@@ -162,6 +162,9 @@ def take_along_axis(arr, indices, axis):
...
```
Comment (mattip) [inline numpy/lib/_shape_base_impl.py:167]:
```suggestion
                'when axis=None, `indices` must have a single dimension.')
```

File: numpy/lib/tests/test_shape_base.py [hunk 4]
```diff
@@ -104,6 +106,21 @@ def test_broadcast(self):
...
+
+        # invalid indices
+        a = a_base.copy()
+        assert_raises(ValueError, put_along_axis, a, indices, values, axis=None)
```
Comment (mattip) [inline numpy/lib/tests/test_shape_base.py:121]: This test will also pass before the PR. 
```suggestion
        with assert_raises(ValueError) as exc:
            put_along_axis(a, indices, values, axis=None)
        assert "single dimension" in str(exc.exception)
```

**Review by mattip — COMMENTED** [review]
Thanks. A couple of suggestions, feel free to reject if you prefer your version

**Comment by arunppsg** [comment]
Thanks for the feedback, I have incorporated the suggestions in ba324c0.

Also, a question: what linter should I use to format to numpy code?

**Comment by mattip** [comment]
Thanks @arunppsg. As for linters: NumPy has a pre-commit hook that (among other things) does some linting, so I think you could use that.

**Pull request merged**